### ISSUE TITLE
Including Product or Price Id's in Payment Document

### ISSUE THREAD

**Issue by TrillyWonka** [stripe/stripe-firebase-extensions#181]
# Feature request

- Extension name: firestore-stripe-subscriptions-handleWebhookEvents

## Is your feature request related to a problem? Please describe.

I've started using the one-time-payments feature with checkout sessions. When the checkout session is completed the webhook fires and Firestore is updated with a new payment document. The issue I'm having is that the payment document makes no reference to the item, product, or price. This means I can't figure out what the customer has paid for (without going into Stripe and checking it manually).

The checkout session document that gets created includes this information, but as far as I can tell there is no way to link a specific payment document to a specific checkout session document becuase there is no overlapping data between them.

I can see the payment object that stripe sends via the webhook contains information about the checkout session, which could then be used to link the checkout session document to the payment document in Firestore, however this data is not captured by Firestore when it arrives.

## Describe the solution you'd like

From what I can see the easiest way to solve this problem would be to have the payment document reference the checkout-id, which is being sent by the Stripe webhook anyway. This could then be used to link the two and I could get the product information from the checkout session focument.

**Comment by thorsten-stripe** [comment]
That's true, that would be useful. I'll look into it.

**Comment by TrillyWonka** [comment]
Thanks Thorsten, much appreciated.

**Comment by rishabhjain** [comment]
Is payment document only being created on live mode? I'm using stripe test mode and can see `checkout_sessions` collection being updated with new documents, not payment collections and docs.

**Comment by rishabhjain** [comment]
Yeah, I have added this to webhook events, but don't see payment document. How are you confirming successful payment from user?

**Comment by japser36** [comment]
This enhancement is exactly what we would need for our use case too!

On our system, we let users subscribe to plans, and this extension works really well and has saved us a lot of time for that. But, we also allow users to purchase extra one time 'topups' if they need something extra in our system on top of what they already get from a subscription, if for example one month is really busy for one of our users.

Without the product information or any reference to the checkout session, there doesn't seem to be a way to support this via the extension without adding a custom webhook to handle those events related to one time payments.

We tried adding metadata to the checkout session creation, since in the extension docs it says this metadata should sync with the subscription object and we were hoping it would also get attached to the payment document, but in our testing this did not work. THe payment document generated with only empty metadata {}.

If this could be added it would really help us out!

**Comment by japser36** [comment]
For now, we worked around this by undeploying the createCheckoutSession extension cloud function, and deploying our own version based on the source in this repo but with a modified 5 lines:
```diff
		if (mode === 'subscription') {
			sessionCreateParams.subscription_data = {
				trial_from_plan,
				metadata,
				default_tax_rates: tax_rates,
			};
+		} else if (mode === 'payment') {
+			sessionCreateParams.payment_intent_data = {
+				metadata,
+			};
		}
```
Then, we passed in the ID of each product the user purchased (we only needed product level info, not price level info for our use case), joined by a ',' which we then parsed in our `/<customer-collection>/{uid}/payments/{payid}` listener, fetching the product documents to get the data to trigger the side effects we needed.

**Comment by MitchelSt** [comment]
> For now, we worked around this by undeploying the createCheckoutSession extension cloud function, and deploying our own version based on the source in this repo but with an added 4 lines:
> 
> ```diff
> 		if (mode === 'subscription') {
> 			sessionCreateParams.subscription_data = {
> 				trial_from_plan,
> 				metadata,
> 				default_tax_rates: tax_rates,
> 			};
> +		} else if (mode === 'payment') {
> +			sessionCreateParams.payment_intent_data = {
> +				metadata,
> +			};
> 		}
> ```
> 
> Then, we passed in the ID of each product the user purchased (we only needed product level info, not price level info for our use case), joined by a ',' which we then parsed in our `/<customer-collection>/{uid}/payments/{payid}` listener, fetching the product documents to get the data to trigger the side effects we needed.
> 
> Unfortunately this does mean we would have to re-undeploy the extension trigger, update and re-deploy our modified version every time we wish to reconfigure something about the firebase extension, but this works for now as a workaround. It would be awesome if this change could be included in the extension!

Adding the snippet above will indeed add the `metadata` for me to the checkout session document. However, I'm not sure about the last part of your implementation.

I assume (and the same goes for me) you want to ensure that the `metadata` of the checkout session is available in it's related payment document as well, isn't it? If yes: may I ask how you did that?

**Comment by thorsten-stripe** [comment]
This will be fixed in v0.1.13

**Comment by japser36** [comment]
> 
> 
> > For now, we worked around this by undeploying the createCheckoutSession extension cloud function, and deploying our own version based on the source in this repo but with an added 4 lines:
> > ```diff
> > 		if (mode === 'subscription') {
> > 			sessionCreateParams.subscription_data = {
> > 				trial_from_plan,
> > 				metadata,
> > 				default_tax_rates: tax_rates,
> > 			};
> > +		} else if (mode === 'payment') {
> > +			sessionCreateParams.payment_intent_data = {
> > +				metadata,
> > +			};
> > 		}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then, we passed in the ID of each product the user purchased (we only needed product level info, not price level info for our use case), joined by a ',' which we then parsed in our `/<customer-collection>/{uid}/payments/{payid}` listener, fetching the product documents to get the data to trigger the side effects we needed.
> > Unfortunately this does mean we would have to re-undeploy the extension trigger, update and re-deploy our modified version every time we wish to reconfigure something about the firebase extension, but this works for now as a workaround. It would be awesome if this change could be included in the extension!
> 
> Adding the snippet above will indeed add the `metadata` for me to the checkout session document. However, I'm not sure about the last part of your implementation.
> 
> I assume (and the same goes for me) you want to ensure that the `metadata` of the checkout session is available in it's related payment document as well, isn't it? If yes: may I ask how you did that?

For us adding that snippet caused `metadata` from the checkout session to be written to the payment document's `metadata` property. The last part of our implementation was basically just using this metadata to write down which products were included in the checkout session, which we consumed in a DB trigger on the payment document write to run some side effects (updating other aspects of our user in our DB on successful payment, which required knowing what products were paid for).

But it looks like they added the products and prices directly :) So for our use case, we no longer need the metadata of the checkout session, we can just read the prices (and thus the products) directly from the payment document now. Thanks @thorsten-stripe !
